### PR TITLE
Fixed BaseSensorOperator to make respect the trigger_rule in downstream tasks, when setting soft_fail="True"

### PR DESCRIPTION
**Pull request by j-y-matsubara**
**Description**
This PR is to prevents all downstream tasks from being forcibly skipped when setting `soft_fail=True`.  

In the current specification, the downstream task will show up as skipped because its trigger_rule is set to all_success by default and skipped tasks will cascade through all_success. Therefore, It is unlikely that this change will affect the existing behavior.

**Issue link**
#8696 
>Downstream tasks with trigger_rule all_success (i.e. the default) should be skipped because of TriggerRuleDep. Tasks that are none_failed or one_success etc should not be skipped unconditionally by the soft_fail sensor.

**Code quality**
flake8 

---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Comment by jeffolsi** [comment]
This behavior needs to be triple checked. You are assuming that the downstream has only one branch which doesn't join with other branches.
If there is a join there is a question of should the skip cascade or stop at the join.

**Comment by j-y-matsubara** [comment]
Thank you for your comments.

> This behavior needs to be triple checked. You are assuming that the downstream has only one branch which doesn't join with other branches.
> If there is a join there is a question of should the skip cascade or stop at the join.

Everything will be skipped, including downstream tasks of the downstream branches.
( This is the same behavior as when soft_fail is set True in the current BaseSensorOperator. )

**Comment by j-y-matsubara** [comment]
@potiuk @kaxil 
Would you mind confirming this PR?

File: airflow/sensors/base_sensor_operator.py
Comment (yuqian90) [inline airflow/sensors/base_sensor_operator.py]: This logic does not look right. It is changing the original behaviour even when user sets `downstream_skip=True`. We should not skip downstream tasks and then raise `AirflowSensorTimeout`. I can't think of any scenario that will be a useful behaviour. The original behaviour is to just raise `AirflowSensorTimeout`.

File: airflow/sensors/base_sensor_operator.py
Comment (yuqian90) [inline airflow/sensors/base_sensor_operator.py]: Why is the default None? If you want the default to be the original behaviour before this PR, it should default to True and handled accordingly in `_timeout_task()`.

**Review by yuqian90 — CHANGES_REQUESTED** [review]

File: airflow/sensors/base_sensor_operator.py
Comment (j-y-matsubara) [reply airflow/sensors/base_sensor_operator.py]: Thank you your comments, @yuqian90 .

>It is changing the original behaviour even when user sets downstream_skip=True. 
Yes.
When setting downstream_skip=False or not setting downstream_skip, then not changing the original behaviour.

The reason I made this change is that :  
I wasn't sure if "downstream_skip" can be added as an argument that **only works when soft_fail is set to true.**

Therefore, I don't disagree with your opinion.
>I can't think of any scenario that will be a useful behaviour.

I will fixe it.

File: airflow/sensors/base_sensor_operator.py
Comment (j-y-matsubara) [reply airflow/sensors/base_sensor_operator.py]: @yuqian90 

This is because the default behavior is different when soft_fail is not set / set False or when soft_fail=True.
When setting `soft_fail = True` ,  All downstream tasks are skipped forcibly.
When setting `soft_fail = False`,  All downstream tasks are not skipped forcibly.

However, I will fix it according to your opinion.
>The original behaviour is to just raise AirflowSensorTimeout.

Comment (yuqian90) [reply airflow/sensors/base_sensor_operator.py]: Thinking about this, I find it unnecessary to provide the `downstream_skip` option. We should just remove the call to `_do_skip_downstream_tasks()` altogether. That can be the only behaviour of `soft_fail`. 

My reason is that whoever not finding `_do_skip_downstream_tasks()` counter-intuitive is most likely using only `all_success` trigger rule in their DAGs. If we remove `_do_skip_downstream_tasks()`, those DAGs will not be affected. Things will continue to work the same. That said, you should still highlight this in the CHANGELOG so people are aware.

Comment (j-y-matsubara) [reply airflow/sensors/base_sensor_operator.py]: @yuqian90 
 I agree with you.
>We should just remove the call to _do_skip_downstream_tasks() altogether. 

Indeed, when all_success trigger_rule is set, the downstream task is automatically skipped when the upstream task is skipped.

I will fix it!

**Comment by j-y-matsubara** [comment]
@yuqian90 
I fixed.
Would you mind confirm this PR, again?

On a side note...
If you know of a documentation file that needs to be modified, please let me know.

**Comment by yuqian90** [comment]
> On a side note...
> If you know of a documentation file that needs to be modified, please let me know.

I think the document you need to update is UPDATING.md. It should mention something about the removal of `_do_skip_downstream_tasks()` and suggest how people can preserve/achieve the original behaviour (by setting every downstream task to `all_success`).

Also pls ping a committer on slack to get them to review/merge. Or try posting in the #development channel.

**Comment by j-y-matsubara** [comment]
> > On a side note...
> > If you know of a documentation file that needs to be modified, please let me know.
> 
> I think the document you need to update is UPDATING.md. It should mention something about the removal of `_do_skip_downstream_tasks()` and suggest how people can preserve/achieve the original behaviour (by setting every downstream task to `all_success`).
> 
> Also pls ping a committer on slack to get them to review/merge. Or try posting in the #development channel.

I appreciate your cooperation. 
Your opinion was very helpful for me.

**Comment by yuqian90** [comment]
Can you look into "continuous-integration/travis-ci/pr — The Travis CI build failed" ?
The build should not be failing. Try rebase onto the latest origin master and push again first to pick up changes by others.

File: UPDATING.md
Comment (yuqian90) [inline UPDATING.md]:
```suggestion
### Made BaseSensorOperator respect the trigger_rule of downstream tasks
```

File: UPDATING.md
Comment (yuqian90) [inline UPDATING.md]:
```suggestion
The point is it does not respect the trigger_rule of downstream tasks when it fails.
```

**Comment by yuqian90** [comment]
Thanks. I commented about some small fixes to the doc inline. Please investigate the `continuous-integration/travis-ci/pr` failure and I'll accept once it passes.

**Comment by j-y-matsubara** [comment]
@yuqian90 
Thank you your advise.
I have corrected the document that you pointed out.

And..
The continuous-integration/travis-ci/pr failure doesn't seem to have anything to do with this PR.
In fact, a similar (following ) travis error has occurred in other pull requests.
There seems to be a problem with the CI environment right now.　
#8919 
I will wait for these to be resolved.

```
Finished the script ci_run_airflow_testing.sh
Elapsed time spent in the script: 1037 seconds
Exit code 1
The command "./scripts/ci/ci_run_airflow_testing.sh" exited with 1.
```

File: UPDATING.md
Comment (j-y-matsubara) [reply UPDATING.md]: Fixed!

File: UPDATING.md
Comment (j-y-matsubara) [reply UPDATING.md]: Fixed!

**Comment by j-y-matsubara** [comment]
@yuqian90 
All tests have passed.
Would you mind confirming this PR again?

**Review by yuqian90 — APPROVED** [review]
Great thanks! Pls chat with a committer on slack to ask them to help you merge this.

**Comment by dossett** [comment]
I agree with this change completely!  I raised this issue on the dev mailing list [here](https://lists.apache.org/thread.html/r1658c78e0bf49d6cdfdd4794044147c69d7874534e45d63c0335a6dc%40%3Cdev.airflow.apache.org%3E) before I knew about this PR.  Thank you @yuqian90 for mentioning it.

**Comment by feluelle** [comment]
@j-y-matsubara can you rebase your commits on master to see if solves the failing checks? They look unrelated to me.

**Comment by j-y-matsubara** [comment]
@feluelle 
> @j-y-matsubara can you rebase your commits on master to see if solves the failing checks? They look unrelated to me.
Sure.

All tests were passed.

File: UPDATING.md
Comment (feluelle) [inline UPDATING.md]: Can you move that to the top when you rebase? :)

**Review by feluelle — COMMENTED** [review]
Besides a minor issue, LGTM 👍

**Comment by feluelle** [comment]
Do we need to update the docs somewhere? Did you check that?

File: UPDATING.md
Comment (j-y-matsubara) [reply UPDATING.md]: I fixed it. :)

**Comment by j-y-matsubara** [comment]
I don't think there is any document that needs to be modified, at least in [airflow/docs](https://github.com/apache/airflow/tree/master/docs).

Is the content of this web page [ (Airflow webpage : base_sensor_operator ) ](https://airflow.apache.org/docs/stable/_api/airflow/sensors/base_sensor_operator/index.html#airflow.sensors.base_sensor_operator.BaseSensorOperator) dynamically created?  
If not ( static page ), editing is required.  But... I can not find the file to be fixed.

**Comment by j-y-matsubara** [comment]
Fixed conflicting!

**Review by feluelle — APPROVED** [review]

File: UPDATING.md
Comment (feluelle) [inline UPDATING.md]: Sorry that I am bothering you again with that, but the UPDATING.md changed a lot i.e. we now have sections.
More information: https://github.com/apache/airflow/pull/10090

Please rebase and move that into an appropriate section.

After you've done that let me know and I am merging it. Sorry and thanks for your patience :)

**Comment by j-y-matsubara** [comment]
Fix conflicting!

File: UPDATING.md
Comment (j-y-matsubara) [reply UPDATING.md]: Thank you for sharing your information with me!
I've fixed! :)

**Review by feluelle — APPROVED** [review]

**Comment by matt-martin** [comment]
I see this was merged, but not sure what release this change would've gone out in. Does anybody know? Also should [AIRFLOW-6163](https://issues.apache.org/jira/browse/AIRFLOW-6163) be marked as resolved?

**Comment by potiuk** [comment]
You need to look at the changelog of Airflow. You wil find it in in airflow.apache.org. ANd no.  First of all JIRA is not used for more than 3 years it is put in read-only state and we do not use it any more, also in our changelog we use PR references not issue references. You can see the PR# and find a reference in the Changelog.

**Pull request merged**
